File: jar/src/main/kotlin/jline/examples/java/basic/PrioModel.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */

package jline.examples.java.basic;

import jline.lang.ClosedClass;
import jline.lang.Network;
import jline.lang.OpenClass;
import jline.lang.RoutingMatrix;
import jline.lang.constant.SchedStrategy;
import jline.lang.nodes.Delay;
import jline.lang.nodes.Queue;
import jline.lang.nodes.Sink;
import jline.lang.nodes.Source;
import jline.lang.processes.Erlang;
import jline.lang.processes.Exp;
import jline.lang.processes.HyperExp;
import jline.solvers.SolverOptions;
import jline.solvers.jmt.JMTOptions;
import jline.solvers.ssa.SSA;
import org.jetbrains.annotations.NotNull;

/**
 * Examples of queueing models with priorities
 */
public class PrioModel {

    /**
     * Open network with multiple scheduling strategies and priority classes.
     * <p>
     * Features:
     * - Three open classes with different priority levels (Class2 has priority 1)
     * - Five stations with different scheduling: FCFS, SIRO, PS, HOL
     * - HOL (Head-of-Line) queue demonstrates priority scheduling
     * - Complex routing with 25% probability splits from central queue
     * - All classes have identical arrival rates but different service rates
     * - Demonstrates impact of scheduling strategies on class performance
     *
     * @return configured priority-based open network model
     */
    public static Network prio_hol_open() {
        Network model = new Network("MyNetwork");

        // Block 1: nodes
        Source source = new Source(model, "Source");
        Queue queue1 = new Queue(model, "WebServer", SchedStrategy.FCFS);
        Queue queue2 = new Queue(model, "Storage1", SchedStrategy.SIRO);
        Queue queue3 = new Queue(model, "Storage2", SchedStrategy.PS);
        Queue queue4 = new Queue(model, "Storage3", SchedStrategy.HOL);
        Sink sink = new Sink(model, "Sink");

        // Block 2: classes
        OpenClass jobclass1 = new OpenClass(model, "Class1", 0);
        OpenClass jobclass2 = new OpenClass(model, "Class2", 1);
        OpenClass jobclass3 = new OpenClass(model, "Class3", 0);

        source.setArrival(jobclass1, Exp.fitMean(10.00)); // (Source,Class1)
        source.setArrival(jobclass2, Exp.fitMean(10.00)); // (Source,Class2)
        source.setArrival(jobclass3, Exp.fitMean(10.00)); // (Source,Class3)
        queue1.setService(jobclass1, Exp.fitMean(0.300)); // (Queue1,Class1)
        queue1.setService(jobclass2, Exp.fitMean(0.500)); // (Queue1,Class2)
        queue1.setService(jobclass3, Exp.fitMean(0.600)); // (Queue1,Class3)
        queue2.setService(jobclass1, Exp.fitMean(1.100)); // (Queue2,Class1)
        queue2.setService(jobclass2, Exp.fitMean(1.300)); // (Queue2,Class2)
        queue2.setService(jobclass3, Exp.fitMean(1.500)); // (Queue2,Class3)
        queue3.setService(jobclass1, Exp.fitMean(2.00)); // (Queue3,Class1)
        queue3.setService(jobclass2, Exp.fitMean(2.100)); // (Queue3,Class2)
        queue3.setService(jobclass3, Exp.fitMean(1.900)); // (Queue3,Class3)
        queue4.setService(jobclass1, Exp.fitMean(2.500)); // (Queue4,Class1)
        queue4.setService(jobclass2, Exp.fitMean(1.900)); // (Queue4,Class2)
        queue4.setService(jobclass3, Exp.fitMean(4.300)); // (Queue4,Class3)

        // Block 3: topology
        RoutingMatrix routingMatrix = model.initRoutingMatrix();

        routingMatrix.set(jobclass1, jobclass1, source, queue1, 1.00); // (Source,Class1) -> (Queue1,Class1)

        routingMatrix.set(jobclass1, jobclass1, queue1, queue2, 0.250000); // (Queue1,Class1) -> (Queue2,Class1)
        routingMatrix.set(jobclass1, jobclass1, queue1, queue3, 0.250000); // (Queue1,Class1) -> (Queue3,Class1)
        routingMatrix.set(jobclass1, jobclass1, queue1, queue4, 0.250000); // (Queue1,Class1) -> (Queue4,Class1)
        routingMatrix.set(jobclass1, jobclass1, queue1, sink, 0.250000); // (Queue1,Class1) -> (Sink,Class1)

        routingMatrix.set(jobclass1, jobclass1, queue2, queue1, 1.00); // (Queue2,Class1) -> (Queue1,Class1)
        routingMatrix.set(jobclass1, jobclass1, queue3, queue1, 1.00); // (Queue3,Class1) -> (Queue1,Class1)
        routingMatrix.set(jobclass1, jobclass1, queue4, queue1, 1.00); // (Queue4,Class1) -> (Queue1,Class1)
        routingMatrix.set(jobclass2, jobclass2, source, queue1, 1.00); // (Source,Class2) -> (Queue1,Class2)

        routingMatrix.set(jobclass2, jobclass2, queue1, queue2, 0.250000); // (Queue1,Class2) -> (Queue2,Class2)
        routingMatrix.set(jobclass2, jobclass2, queue1, queue3, 0.250000); // (Queue1,Class2) -> (Queue3,Class2)
        routingMatrix.set(jobclass2, jobclass2, queue1, queue4, 0.250000); // (Queue1,Class2) -> (Queue4,Class2)
        routingMatrix.set(jobclass2, jobclass2, queue1, sink, 0.250000); // (Queue1,Class2) -> (Sink,Class2)

        routingMatrix.set(jobclass2, jobclass2, queue2, queue1, 1.00); // (Queue2,Class2) -> (Queue1,Class2)
        routingMatrix.set(jobclass2, jobclass2, queue3, queue1, 1.00); // (Queue3,Class2) -> (Queue1,Class2)
        routingMatrix.set(jobclass2, jobclass2, queue4, queue1, 1.00); // (Queue4,Class2) -> (Queue1,Class2)
        routingMatrix.set(jobclass3, jobclass3, source, queue1, 1.00); // (Source,Class3) -> (Queue1,Class3)

        routingMatrix.set(jobclass3, jobclass3, queue1, queue2, 0.250000); // (Queue1,Class3) -> (Queue2,Class3)
        routingMatrix.set(jobclass3, jobclass3, queue1, queue3, 0.250000); // (Queue1,Class3) -> (Queue3,Class3)
        routingMatrix.set(jobclass3, jobclass3, queue1, queue4, 0.250000); // (Queue1,Class3) -> (Queue4,Class3)
        routingMatrix.set(jobclass3, jobclass3, queue1, sink, 0.250000); // (Queue1,Class3) -> (Sink,Class3)

        routingMatrix.set(jobclass3, jobclass3, queue2, queue1, 1.00); // (Queue2,Class3) -> (Queue1,Class3)
        routingMatrix.set(jobclass3, jobclass3, queue3, queue1, 1.00); // (Queue3,Class3) -> (Queue1,Class3)
        routingMatrix.set(jobclass3, jobclass3, queue4, queue1, 1.00); // (Queue4,Class3) -> (Queue1,Class3)

        model.link(routingMatrix);

        return model;
    }

    /**
     * Closed network comparing scheduling strategies with priorities.
     * <p>
     * Features:
     * - Three closed classes (18 jobs each) with Class2 having priority 1
     * - Six stations: SlowDelay, FCFS, SIRO, PS, HOL queues, FastDelay
     * - Central FCFS queue distributes traffic (25% each direction)
     * - HOL queue provides priority service for Class2
     * - Cyclic topology demonstrating sustained priority effects
     * - Identical service rates at delay stations for fair comparison
     *
     * @return configured closed priority network model
     */
    public static Network prio_hol_closed() {
        Network model = new Network("MyNetwork");

        // Block 1: nodes
        Delay node1 = new Delay(model, "SlowDelay");
        Queue node2 = new Queue(model, "FCFSQueue", SchedStrategy.FCFS);
        Queue node3 = new Queue(model, "SIROQueue", SchedStrategy.SIRO);
        Queue node4 = new Queue(model, "PSQueue", SchedStrategy.PS);
        Queue node5 = new Queue(model, "HOLQueue", SchedStrategy.HOL);
        Delay node6 = new Delay(model, "FastDelay");

        // Block 2: classes
        ClosedClass jobclass1 = new ClosedClass(model, "Class1", 18, node1, 0);
        ClosedClass jobclass2 = new ClosedClass(model, "Class2", 18, node1, 1);
        ClosedClass jobclass3 = new ClosedClass(model, "Class3", 18, node1, 0);

        node1.setService(jobclass1, Exp.fitMean(10.00)); // (SlowDelay,Class1)
        node1.setService(jobclass2, Exp.fitMean(10.00)); // (SlowDelay,Class2)
        node1.setService(jobclass3, Exp.fitMean(10.00)); // (SlowDelay,Class3)

        node2.setService(jobclass1, Exp.fitMean(0.300)); // (FCFSQueue,Class1)
        node2.setService(jobclass2, Exp.fitMean(0.500)); // (FCFSQueue,Class2)
        node2.setService(jobclass3, Exp.fitMean(0.600)); // (FCFSQueue,Class3)

        node3.setService(jobclass1, Exp.fitMean(1.100)); // (SIROQueue,Class1)
        node3.setService(jobclass2, Exp.fitMean(1.300)); // (SIROQueue,Class2)
        node3.setService(jobclass3, Exp.fitMean(1.500)); // (SIROQueue,Class3)

        node4.setService(jobclass1, Exp.fitMean(1.00)); // (PSQueue,Class1)
        node4.setService(jobclass2, Exp.fitMean(1.100)); // (PSQueue,Class2)
        node4.setService(jobclass3, Exp.fitMean(1.900)); // (PSQueue,Class3)

        node5.setService(jobclass1, Exp.fitMean(2.500)); // (HOLQueue,Class1)
        node5.setService(jobclass2, Exp.fitMean(1.900)); // (HOLQueue,Class2)
        node5.setService(jobclass3, Exp.fitMean(4.300)); // (HOLQueue,Class3)

        node6.setService(jobclass1, Exp.fitMean(1.00)); // (FastDelay,Class1)
        node6.setService(jobclass2, Exp.fitMean(1.00)); // (FastDelay,Class2)
        node6.setService(jobclass3, Exp.fitMean(1.00)); // (FastDelay,Class3)

        // Block 3: topology
        RoutingMatrix routingMatrix = model.initRoutingMatrix();

        routingMatrix.set(jobclass1, jobclass1, node1, node2, 1.00); // (SlowDelay,Class1) -> (FCFSQueue,Class1)
        routingMatrix.set(jobclass1, jobclass1, node2, node3, 0.250000); // (FCFSQueue,Class1) -> (SIROQueue,Class1)
        routingMatrix.set(jobclass1, jobclass1, node2, node4, 0.250000); // (FCFSQueue,Class1) -> (PSQueue,Class1)
        routingMatrix.set(jobclass1, jobclass1, node2, node5, 0.250000); // (FCFSQueue,Class1) -> (HOLQueue,Class1)
        routingMatrix.set(jobclass1, jobclass1, node2, node6, 0.250000); // (FCFSQueue,Class1) -> (FastDelay,Class1)

        routingMatrix.set(jobclass1, jobclass1, node3, node2, 1.00); // (SIROQueue,Class1) -> (FCFSQueue,Class1)
        routingMatrix.set(jobclass1, jobclass1, node4, node2, 1.00); // (PSQueue,Class1) -> (FCFSQueue,Class1)
        routingMatrix.set(jobclass1, jobclass1, node5, node2, 1.00); // (HOLQueue,Class1) -> (FCFSQueue,Class1)

        routingMatrix.set(jobclass2, jobclass2, node1, node2, 1.00); // (SlowDelay,Class2) -> (FCFSQueue,Class2)
        routingMatrix.set(jobclass2, jobclass2, node2, node3, 0.250000); // (FCFSQueue,Class2) -> (SIROQueue,Class2)
        routingMatrix.set(jobclass2, jobclass2, node2, node4, 0.250000); // (FCFSQueue,Class2) -> (PSQueue,Class2)
        routingMatrix.set(jobclass2, jobclass2, node2, node5, 0.250000); // (FCFSQueue,Class2) -> (HOLQueue,Class2)
        routingMatrix.set(jobclass2, jobclass2, node2, node6, 0.250000); // (FCFSQueue,Class2) -> (FastDelay,Class2)

        routingMatrix.set(jobclass2, jobclass2, node3, node2, 1.00); // (SIROQueue,Class2) -> (FCFSQueue,Class2)
        routingMatrix.set(jobclass2, jobclass2, node4, node2, 1.00); // (PSQueue,Class2) -> (FCFSQueue,Class2)
        routingMatrix.set(jobclass2, jobclass2, node5, node2, 1.00); // (HOLQueue,Class2) -> (FCFSQueue,Class2)

        routingMatrix.set(jobclass3, jobclass3, node1, node2, 1.00); // (SlowDelay,Class3) -> (FCFSQueue,Class3)
        routingMatrix.set(jobclass3, jobclass3, node2, node3, 0.250000); // (FCFSQueue,Class3) -> (SIROQueue,Class3)
        routingMatrix.set(jobclass3, jobclass3, node2, node4, 0.250000); // (FCFSQueue,Class3) -> (PSQueue,Class3)
        routingMatrix.set(jobclass3, jobclass3, node2, node5, 0.250000); // (FCFSQueue,Class3) -> (HOLQueue,Class3)
        routingMatrix.set(jobclass3, jobclass3, node2, node6, 0.250000); // (FCFSQueue,Class3) -> (FastDelay,Class3)

        routingMatrix.set(jobclass3, jobclass3, node3, node2, 1.00); // (SIROQueue,Class3) -> (FCFSQueue,Class3)
        routingMatrix.set(jobclass3, jobclass3, node4, node2, 1.00); // (PSQueue,Class3) -> (FCFSQueue,Class3)
        routingMatrix.set(jobclass3, jobclass3, node5, node2, 1.00); // (HOLQueue,Class3) -> (FCFSQueue,Class3)

        routingMatrix.set(jobclass1, jobclass1, node6, node1, 1.00); // (FastDelay,Class1) -> (SlowDelay,Class1)
        routingMatrix.set(jobclass2, jobclass2, node6, node1, 1.00); // (FastDelay,Class2) -> (SlowDelay,Class2)
        routingMatrix.set(jobclass3, jobclass3, node6, node1, 1.00); // (FastDelay,Class3) -> (SlowDelay,Class3)

        model.link(routingMatrix);

        return model;
    }

    /**
     * Simple priority network with PSPRIO scheduling.
     * <p>
     * Features:
     * - Two closed classes (2 jobs each) with Class2 having priority 1
     * - PSPRIO (Processor Sharing with Priorities) scheduling
     * - Mixed service distributions: Erlang for Class1, HyperExp for Class2
     * - Serial two-station topology: Delay → PSPRIO Queue → Delay
     * - Demonstrates preemptive priority in processor sharing
     *
     * @return configured PSPRIO network model
     */
    @NotNull
    public static Network prio_psprio() {
        Network model = new Network("MyNetwork");

        // Block 1: nodes
        Delay node1 = new Delay(model, "SlowDelay");
        Queue node2 = new Queue(model, "PSPRIOQueue", SchedStrategy.PSPRIO);

        // Block 2: classes
        ClosedClass jobclass1 = new ClosedClass(model, "Class1", 2, node1, 0);
        ClosedClass jobclass2 = new ClosedClass(model, "Class2", 2, node1, 1);

        node1.setService(jobclass1, new Erlang(3.0, 2)); // (SlowDelay,Class1)
        node1.setService(jobclass2, new HyperExp(0.5, 3.0, 10.0)); // (SlowDelay,Class2)

        node2.setService(jobclass1, new HyperExp(0.1, 1.0, 10.0)); // (FCFSQueue,Class1)
        node2.setService(jobclass2, new Exp(1)); // (FCFSQueue,Class2)

        // Block 3: topology
        RoutingMatrix routingMatrix = model.initRoutingMatrix();
        routingMatrix.set(jobclass1, Network.serialRouting(node1, node2));
        routingMatrix.set(jobclass2, Network.serialRouting(node1, node2));
        model.link(routingMatrix);
        return model;
    }

    /**
     * Network with GPSPRIO scheduling and multiple priority classes.
     * <p>
     * Features:
     * - Four closed classes with different populations and priorities
     * - Two stations: Delay node and GPSPRIO queue with weights
     * - Class1: 6 jobs, priority 0; Class2,3: 4 jobs, priority 1; Class4: 1 job, priority 2
     * - Demonstrates GPS (Generalized Processor Sharing) with priorities
     * - Different service rates and weights for weighted fair sharing
     * - Simple serial routing: Delay → GPSPRIO Queue → Delay
     *
     * @return configured network with GPSPRIO scheduling
     */
    public static Network prio_identical() {
        Network model = new Network("model");

        // Block 1: nodes
        Delay node1 = new Delay(model, "Delay");
        Queue node2 = new Queue(model, "Queue1", SchedStrategy.GPSPRIO);

        // Block 2: classes with different populations and priorities
        ClosedClass jobclass1 = new ClosedClass(model, "Class1", 6, node1, 0);
        ClosedClass jobclass2 = new ClosedClass(model, "Class2", 4, node1, 1);
        ClosedClass jobclass3 = new ClosedClass(model, "Class3", 4, node1, 1);
        ClosedClass jobclass4 = new ClosedClass(model, "Class4", 1, node1, 2);

        // Service rates at delay station
        node1.setService(jobclass1, new Erlang(3, 2));
        node1.setService(jobclass2, new Exp(1));
        node1.setService(jobclass3, new Exp(1.0));
        node1.setService(jobclass4, new Exp(2.0));

        // GPSPRIO queue with weights and service rates
        int w1 = 12; node2.setService(jobclass1, new Exp(30), w1);
        int w2 = 3;  node2.setService(jobclass2, new Exp(2), w2);
        int w3 = 5;  node2.setService(jobclass3, new Exp(12), w3);
        int w4 = 1;  node2.setService(jobclass4, new Exp(1), w4);

        // Block 3: topology - simple serial routing
        RoutingMatrix routingMatrix = model.initRoutingMatrix();
        
        routingMatrix.set(jobclass1, Network.serialRouting(node1, node2));
        routingMatrix.set(jobclass2, Network.serialRouting(node1, node2));
        routingMatrix.set(jobclass3, Network.serialRouting(node1, node2));
        routingMatrix.set(jobclass4, Network.serialRouting(node1, node2));

        model.link(routingMatrix);
        return model;
    }

    /**
     * Main method for testing and demonstrating priority examples.
     *
     * <p>Currently configured to:
     * - Run prio_psprio() with PSPRIO scheduling
     * - Use JMT options with specified seed and sample count
     * - Solve using SSA solver and print average metrics
     * - Model visualization and JMT solving are commented out
     *
     * @param args command line arguments (not used)
     * @throws Exception if solver encounters an error
     */
    public static void main(String[] args) throws Exception {
        Network model = prio_psprio();
        SolverOptions options = new JMTOptions();
        options.seed = 23000;
        options.samples = 10000;
        //model.view();
        //new JMT(model, options).getAvgTable().print();
        new SSA(model, options).getAvgTable().print();
    }

}

File: cpp/examples/basic/prioModel.cpp
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */

/**
 * `python/examples/basic/prioModel/`: the four class-priority models.
 *
 * Each one puts a priority discipline -- HOL, PSPRIO, GPSPRIO -- beside at least
 * one exact or simulated answer for the same network, so the priority
 * approximation can be read against something. JMT and LDES have no C++ port and
 * are refused by name; CTMC, MVA and SSA are the reference's own blocks and run.
 *
 * WHAT A PRIORITY IS HERE. The class priority is the `prio` argument of
 * `add_open_class` / `add_closed_class` -- a LARGER number is a HIGHER priority,
 * as in `ClosedClass(model, name, N, refnode, prio)` -- and the station decides
 * what to do with it through its scheduling strategy. A class carrying a
 * priority at a station whose discipline ignores one (FCFS, SIRO, PS) is not an
 * error and changes nothing there, which is exactly what `prio_hol_open` and
 * `prio_hol_closed` are built to show.
 */

#include <cstddef>
#include <vector>

#include "example_util.h"
#include "examples_common.h"
#include "line/lang/dist_fitters.h"
#include "line/solvers/ctmc/solver_ctmc_waitq.h"
#include "line/solvers/mva/solver_mva_runner.h"
#include "line/solvers/ssa/ssa_dispatch.h"

namespace line {
namespace examples {

namespace {

/** The reason every JMT block in this directory carries. */
const char* kNoJmt =
    "SolverJMT drives the Java Modelling Tools simulator, which has no C++ port; the native "
    "blocks above answer the same model, and reporting one of them under the simulator's name "
    "would attribute a number to an engine that never produced it";

/** The reason `prio_hol_closed` refuses its fourth block. */
const char* kNoLdes =
    "SolverLDES is the SSJ-based discrete-event engine of the JAR, driven over model.json by the "
    "MATLAB and Python clients; the C++ port carries no LDES engine and no client for it, and SSA "
    "above is a different simulator, not this one";

mva::AvgResult<double> mva_run(Net& m, double iter_tol = -1.0) {
    mva::MvaOptions opt;
    if (iter_tol > 0.0) opt.iter_tol = iter_tol;
    Matrix<double> init;
    return mva::solver_mva_run_analyzer(m.get_struct(), opt, init);
}

mva::AvgResult<double> ctmc_run(Net& m, double cutoff = -1.0) {
    ctmc::CtmcOptions opt;
    opt.cutoff = cutoff;
    return ctmc::solver_ctmc_run_analyzer_any(m.get_struct(), opt);
}

ssa::SsaSolution ssa_run(Net& m, std::size_t samples, unsigned long seed) {
    ssa::SsaOptions opt;
    opt.samples = samples;
    opt.seed = seed;
    return ssa::solver_ssa(m.get_struct(), opt);
}

// ---------------------------------------------------------------------------
// Models
// ---------------------------------------------------------------------------

/**
 * The closed six-node network: a slow Delay, four queues under FCFS, SIRO, PS
 * and HOL, and a fast Delay. Class2 carries priority 1, Class1 and Class3
 * priority 0, so only the HOL queue distinguishes them.
 */
Net prio_hol_closed_model() {
    Net m("MyNetwork");
    Delay slow(m, "SlowDelay");
    Queue fcfs(m, "FCFSQueue", SchedStrategy::FCFS);
    Queue siro(m, "SIROQueue", SchedStrategy::SIRO);
    Queue ps(m, "PSQueue", SchedStrategy::PS);
    Queue hol(m, "HOLQueue", SchedStrategy::HOL);
    Delay fast(m, "FastDelay");

    std::vector<std::size_t> cls;
    cls.push_back(m.add_closed_class("Class1", 18.0, slow, 0));
    cls.push_back(m.add_closed_class("Class2", 18.0, slow, 1));
    cls.push_back(m.add_closed_class("Class3", 18.0, slow, 0));

    const double slow_mean[3] = {10.0, 10.0, 10.0};
    const double fcfs_mean[3] = {0.3, 0.5, 0.6};
    const double siro_mean[3] = {1.1, 1.3, 1.5};
    const double ps_mean[3] = {1.0, 1.1, 1.9};
    const double hol_mean[3] = {2.5, 1.9, 4.3};
    const double fast_mean[3] = {1.0, 1.0, 1.0};
    for (std::size_t r = 0; r < 3; ++r) {
        slow.set_service(cls[r], D::exp_mean(slow_mean[r]));
        fcfs.set_service(cls[r], D::exp_mean(fcfs_mean[r]));
        siro.set_service(cls[r], D::exp_mean(siro_mean[r]));
        ps.set_service(cls[r], D::exp_mean(ps_mean[r]));
        hol.set_service(cls[r], D::exp_mean(hol_mean[r]));
        fast.set_service(cls[r], D::exp_mean(fast_mean[r]));
    }

    Routing P;
    for (std::size_t r = 0; r < 3; ++r) {
        P.set(cls[r], cls[r], slow, fcfs, 1.0);
        P.set(cls[r], cls[r], fcfs, siro, 0.25);
        P.set(cls[r], cls[r], fcfs, ps, 0.25);
        P.set(cls[r], cls[r], fcfs, hol, 0.25);
        P.set(cls[r], cls[r], fcfs, fast, 0.25);
        P.set(cls[r], cls[r], siro, fcfs, 1.0);
        P.set(cls[r], cls[r], ps, fcfs, 1.0);
        P.set(cls[r], cls[r], hol, fcfs, 1.0);
        P.set(cls[r], cls[r], fast, slow, 1.0);
    }
    m.link(P);
    return m;
}

/** The open counterpart: a Source, the same four disciplines, and a Sink. */
Net prio_hol_open_model() {
    Net m("MyNetwork");
    Source source(m, "Source");
    Queue web(m, "WebServer", SchedStrategy::FCFS);
    Queue st1(m, "Storage1", SchedStrategy::SIRO);
    Queue st2(m, "Storage2", SchedStrategy::PS);
    Queue st3(m, "Storage3", SchedStrategy::HOL);
    Sink sink(m, "Sink");

    std::vector<std::size_t> cls;
    cls.push_back(m.add_open_class("Class1", 0));
    cls.push_back(m.add_open_class("Class2", 1));
    cls.push_back(m.add_open_class("Class3", 0));

    const double web_mean[3] = {0.3, 0.5, 0.6};
    const double st1_mean[3] = {1.1, 1.3, 1.5};
    const double st2_mean[3] = {2.0, 2.1, 1.9};
    const double st3_mean[3] = {2.5, 1.9, 4.3};
    for (std::size_t r = 0; r < 3; ++r) {
        source.set_arrival(cls[r], D::exp_mean(10.0));
        web.set_service(cls[r], D::exp_mean(web_mean[r]));
        st1.set_service(cls[r], D::exp_mean(st1_mean[r]));
        st2.set_service(cls[r], D::exp_mean(st2_mean[r]));
        st3.set_service(cls[r], D::exp_mean(st3_mean[r]));
    }

    Routing P;
    for (std::size_t r = 0; r < 3; ++r) {
        P.set(cls[r], cls[r], source, web, 1.0);
        P.set(cls[r], cls[r], web, st1, 0.25);
        P.set(cls[r], cls[r], web, st2, 0.25);
        P.set(cls[r], cls[r], web, st3, 0.25);
        P.set(cls[r], cls[r], web, sink, 0.25);
        P.set(cls[r], cls[r], st1, web, 1.0);
        P.set(cls[r], cls[r], st2, web, 1.0);
        P.set(cls[r], cls[r], st3, web, 1.0);
    }
    m.link(P);
    return m;
}

/**
 * GPSPRIO with a TIE: Class2 and Class3 both carry priority 1, so the queue
 * resolves between them by their GPS weights rather than by their priorities.
 */
Net prio_identical_model() {
    Net m("model");
    Delay delay(m, "Delay");
    Queue queue(m, "Queue1", SchedStrategy::GPSPRIO);

    std::vector<std::size_t> cls;
    cls.push_back(m.add_closed_class("Class1", 6.0, delay, 0));
    cls.push_back(m.add_closed_class("Class2", 4.0, delay, 1));
    cls.push_back(m.add_closed_class("Class3", 4.0, delay, 1));
    cls.push_back(m.add_closed_class("Class4", 1.0, delay, 2));

    delay.set_service(cls[0], Erlang(3.0, 2));
    delay.set_service(cls[1], Exp(1.0));
    delay.set_service(cls[2], Exp(1.0));
    delay.set_service(cls[3], Exp(2.0));

    const double rate[4] = {30.0, 2.0, 12.0, 1.0};
    const double weight[4] = {12.0, 3.0, 5.0, 1.0};
    for (std::size_t r = 0; r < 4; ++r) {
        queue.set_service(cls[r], Exp(rate[r]));
        queue.set_sched_param(cls[r], weight[r]);
    }

    Routing P;
    for (std::size_t r = 0; r < 4; ++r) cyclic(P, cls[r], {delay, queue});
    m.link(P);
    return m;
}

/** PSPRIO: processor sharing inside each priority level, strict between them. */
Net prio_psprio_model() {
    Net m("MyNetwork");
    Delay slow(m, "SlowDelay");
    Queue queue(m, "PSPRIOQueue", SchedStrategy::PSPRIO);

    ClosedClass c1(m, "Class1", 2.0, slow, 0);
    ClosedClass c2(m, "Class2", 2.0, slow, 1);

    slow.set_service(c1, Erlang(3.0, 2));
    slow.set_service(c2, HyperExp(0.5, 3.0, 10.0));
    queue.set_service(c1, HyperExp(0.1, 1.0, 10.0));
    queue.set_service(c2, Exp(1.0));

    Routing P;
    cyclic(P, c1, {slow, queue});
    cyclic(P, c2, {slow, queue});
    m.link(P);
    return m;
}

}  // namespace

// ---------------------------------------------------------------------------
// Examples
// ---------------------------------------------------------------------------

void prio_hol_closed() {
    Net m = prio_hol_closed_model();
    section("MVA");
    print_avg(m.get_struct(), mva_run(m));
    section("JMT");
    print_avg(m.get_struct(), jmt_avg(m, sim_opts(23000, 10000)));
    section("SSA");
    print_avg_sim(m.get_struct(), ssa_run(m, 10000, 23000));
    // TODO(cpp): print(LDES(model, seed=23000, samples=10000).avg_table())
    na("LDES", kNoLdes);
}

/**
 * `cutoff=1` on the CTMC block is the reference's own truncation: the chain of
 * an open network is infinite, so its answer is the answer of the truncated
 * chain and is reported as such rather than as the model's.
 */
void prio_hol_open() {
    Net m = prio_hol_open_model();
    section("CTMC");
    print_avg(m.get_struct(), ctmc_run(m, 1.0));
    // The MVA-specific iter_tol (1e-6), not the generic 1e-4: the looser one
    // stopped the AMVA fixed point 2.9e-5 short on Storage3 and split this row
    // from the engines that solve at the solver's own tolerance.
    section("MVA");
    print_avg(m.get_struct(), mva_run(m));
    section("JMT");
    print_avg(m.get_struct(), jmt_avg(m, sim_opts(23000, 10000)));
    section("SSA");
    print_avg_sim(m.get_struct(), ssa_run(m, 10000, 23000));
}

void prio_identical() {
    Net m = prio_identical_model();
    section("CTMC");
    print_avg(m.get_struct(), ctmc_run(m));
    section("JMT");
    print_avg(m.get_struct(), jmt_avg(m, sim_opts(23000, 10000)));
}

void prio_psprio() {
    Net m = prio_psprio_model();
    section("CTMC");
    print_avg(m.get_struct(), ctmc_run(m));
    section("JMT");
    print_avg(m.get_struct(), jmt_avg(m, sim_opts(23000, 5000)));
    section("SSA");
    print_avg_sim(m.get_struct(), ssa_run(m, 5000, 23000));
}

LINE_EXAMPLE("basic/prioModel", prio_hol_closed);
LINE_EXAMPLE("basic/prioModel", prio_hol_open);
LINE_EXAMPLE("basic/prioModel", prio_identical);
LINE_EXAMPLE("basic/prioModel", prio_psprio);

}  // namespace examples
}  // namespace line
